Parallelize tests

Fizzixnerd opened the pull request:
Requires adding dependency annotations still.

Performance impact is relatively significant: tests run ~2x faster on my machine with 8 logical cores.  Where a single test took ~2.5 seconds to run, now 8 are run in parallel at roughly 6-8s per test.

Fizzixnerd committed:
Parallelize tests

Fizzixnerd commented:
Forgot to mention #1924

ranjitjhala commented:
this is awesome, thank you!!!

Fizzixnerd commented:
This is ready for review, (assuming the tests pass!) but I just wanted to comment on what I changed.  I partitioned the test for each group into two sets: `Lib` tests and non-`Lib` tests.  This assumes that any test that needs to be compiled before another one is named `xLiby.hs` where `xx` and `yy` are arbitrary.  (This seems to be the convention in the tests from my view, but I wanted to make sure I wasn't missing any "dependency tests" that do not follow this naming convention)  The non-`Lib` tests are run after the `Lib` tests always, so this should ensure there is no race conditions -- at least for libs that do not depend on other libs!

This is kind of a quick and dirty solution, but it's automatic and seems to work.  As always with parallel programs however, it may be that things will mysteriously break in the future or on someone else's machine (test wise) because of races, so I thought I'd bring it up, so everyone knows what to expect

facundominguez commented on tests/test.hs:
```diff
@@ -534,12 +540,17 @@ mkTest ec dir file
+logFilePath :: FilePath -> FilePath -> FilePath
+logFilePath dir file = "tests/logs/cur" </> testFileFullPath <.> "log"
+  where
+    testFileFullPath = dir </> file
```
This function doesn't appear to be used anywhere.

Fizzixnerd replied:
Removed.

facundominguez commented on tests/test.hs:
```diff
@@ -414,29 +411,35 @@ textOrder = Just . mkOrder $
 -- | For each file in `root` check, that we get the given exit `code.`
 --------------------------------------------------------------------------------
-dirTests :: FilePath -> [FilePath] -> ExitCode -> Maybe T.Text -> Maybe T.Text -> IO [TestTree]
+dirTests :: FilePath -> [FilePath] -> ExitCode -> Maybe T.Text -> Maybe T.Text -> IO ([TestTree], [TestTree])
```
How about making a record type to stand for the libs and the tests?
```haskell
data DependentTests = DependentTests
  { dependencies :: [TestTree]
  , toplevel :: [TestTree]
  }
```
It would help documenting the meaning of the output and other functions when it appears in the parameters.

Fizzixnerd replied:
Added.

facundominguez reviewed (commented):
Thank you, Matt!

> This assumes that any test that needs to be compiled before another one is named xLiby.hs where xx and yy are arbitrary.

How about mentioning this at the place in the code where the names are introduced?

The change looks good to me, though the times in CI do not appear to have reduced as dramatically. @Fizzixnerd, can you have a look? Perhaps some long tests simply don't parallelize well.

ranjitjhala commented:
The xLib idea sounds good to me!

On Mon, Jan 17, 2022 at 4:27 PM Facundo Domínguez ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
>
> Thank you, Matt!
>
> This assumes that any test that needs to be compiled before another one is
> named xLiby.hs where xx and yy are arbitrary.
>
> How about mentioning this at the place in the code where the names are
> introduced?
>
> The change looks good to me, though the times in CI do not appear to have
> reduced as dramatically. @Fizzixnerd
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_Fizzixnerd&d=DwMCaQ&c=-35OiAkTchMrZOngvJPOeA&r=r3JfTqNkpwIJ1InE9-ChC2ld7xwATxgUx5XHAdA0UnA&m=7jf5GFnsY3f5FL0aDDNkJkSwtxJ9SL40rnZqtTrxs-aljy9cQFjmbV1kB9VslUZh&s=G4vZRAhLNjoPmNYvrN68CZVEMsmJ8E4MTeMAGhTMmVM&e=>,
> do you think you could check how many threads are used with -N in CI? (I
> think you can enable CircleCI in your fork of liquidhaskell for this sake)
>
> Perhaps using -N2 or -N4 would make a difference. Though while the
> overall test time doesn't worsen, the investigation outcome doesn't need to
> go in this PR necessarily.
> ------------------------------
>
> In tests/test.hs
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_ucsd-2Dprogsys_liquidhaskell_pull_1925-23discussion-5Fr786338589&d=DwMCaQ&c=-35OiAkTchMrZOngvJPOeA&r=r3JfTqNkpwIJ1InE9-ChC2ld7xwATxgUx5XHAdA0UnA&m=7jf5GFnsY3f5FL0aDDNkJkSwtxJ9SL40rnZqtTrxs-aljy9cQFjmbV1kB9VslUZh&s=NtWotTZA3G55C5kZRT7Z9szYbMS0U-xeM3RZg-ya-lg&e=>
> :
>
> > @@ -534,12 +540,17 @@ mkTest ec dir file
>      test = dir </> file
>
>
> +logFilePath :: FilePath -> FilePath -> FilePath
> +logFilePath dir file = "tests/logs/cur" </> testFileFullPath <.> "log"
> +  where
> +    testFileFullPath = dir </> file
>
> This function doesn't appear to be used anywhere.
> ------------------------------
>
> In tests/test.hs
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_ucsd-2Dprogsys_liquidhaskell_pull_1925-23discussion-5Fr786343452&d=DwMCaQ&c=-35OiAkTchMrZOngvJPOeA&r=r3JfTqNkpwIJ1InE9-ChC2ld7xwATxgUx5XHAdA0UnA&m=7jf5GFnsY3f5FL0aDDNkJkSwtxJ9SL40rnZqtTrxs-aljy9cQFjmbV1kB9VslUZh&s=HYjotx5n6wuYzHkx-eIc8jSbU4UGWFFe93QIWCvOEKg&e=>
> :
>
> >    ]
>
>  --------------------------------------------------------------------------------
>  -- | For each file in `root` check, that we get the given exit `code.`
>  --------------------------------------------------------------------------------
> -dirTests :: FilePath -> [FilePath] -> ExitCode -> Maybe T.Text -> Maybe T.Text -> IO [TestTree]
> +dirTests :: FilePath -> [FilePath] -> ExitCode -> Maybe T.Text -> Maybe T.Text -> IO ([TestTree], [TestTree])
>
> How about making a record type to stand for the libs and the tests?
>
> data DependentTests = DependentTests
>   { dependencies :: [TestTree]
>   , toplevel :: [TestTree]
>   }
>
> It would help documenting the meaning of the output and other functions
> when it appears in the parameters.
>
> —
> Reply to this email directly, view it on GitHub
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_ucsd-2Dprogsys_liquidhaskell_pull_1925-23pullrequestreview-2D854844822&d=DwMCaQ&c=-35OiAkTchMrZOngvJPOeA&r=r3JfTqNkpwIJ1InE9-ChC2ld7xwATxgUx5XHAdA0UnA&m=7jf5GFnsY3f5FL0aDDNkJkSwtxJ9SL40rnZqtTrxs-aljy9cQFjmbV1kB9VslUZh&s=BYR_PbGNEzkLaSyAEgvBzdOTaqHYXLaS4j8meT-mpv0&e=>,
> or unsubscribe
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_notifications_unsubscribe-2Dauth_AAMS4OD6KULZNP4YLKHVM7LUWSXWBANCNFSM5MFHOX3Q&d=DwMCaQ&c=-35OiAkTchMrZOngvJPOeA&r=r3JfTqNkpwIJ1InE9-ChC2ld7xwATxgUx5XHAdA0UnA&m=7jf5GFnsY3f5FL0aDDNkJkSwtxJ9SL40rnZqtTrxs-aljy9cQFjmbV1kB9VslUZh&s=omvLwxqLA_EO_dwmRl4R7q4giGd2l_YlvkwBMm95V30&e=>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__apps.apple.com_app_apple-2Dstore_id1477376905-3Fct-3Dnotification-2Demail-26mt-3D8-26pt-3D524675&d=DwMCaQ&c=-35OiAkTchMrZOngvJPOeA&r=r3JfTqNkpwIJ1InE9-ChC2ld7xwATxgUx5XHAdA0UnA&m=7jf5GFnsY3f5FL0aDDNkJkSwtxJ9SL40rnZqtTrxs-aljy9cQFjmbV1kB9VslUZh&s=zHnRsNlhByYVQBh25fRBzg5WSxFtaaHG34W5odBnCYs&e=>
> or Android
> <https://urldefense.proofpoint.com/v2/url?u=https-3A__play.google.com_store_apps_details-3Fid-3Dcom.github.android-26referrer-3Dutm-5Fcampaign-253Dnotification-2Demail-2526utm-5Fmedium-253Demail-2526utm-5Fsource-253Dgithub&d=DwMCaQ&c=-35OiAkTchMrZOngvJPOeA&r=r3JfTqNkpwIJ1InE9-ChC2ld7xwATxgUx5XHAdA0UnA&m=7jf5GFnsY3f5FL0aDDNkJkSwtxJ9SL40rnZqtTrxs-aljy9cQFjmbV1kB9VslUZh&s=UaduuYp_x1WoQk5qNbZLhWP3LETN_MUUt51v662jQp4&e=>.
>
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

Fizzixnerd committed:
Remove unused function; create new type for dependent tests

Fizzixnerd committed:
Control parallelism for sequential tests

Fizzixnerd commented:
I noticed on my machine some tests sometimes failing in the "big library projects" (ByteString, Text, etc), so I changed anything that has a `FileOrder` to a `SequentialFileOrder` (just a newtype) and added a function that runs `SequentialFileOrder`ed tests in the order they appear in the order.

To summarize, use `testGroupsWithLibs` with `[o]dirTests` when you want to have parallel tests, and use `testSequentially` with `sequentialOdirTests` when you want to force sequential execution of tests.

Fizzixnerd committed:
Document the Lib behavior better

Fizzixnerd commented:
Building on CI takes 13 minutes for stack_810, which would not change with the -N, but cabal_810 is still building dependencies as of 15 minutes in.  I think this is probably the reason for the less than-impressive speed up -- much of the time in CI is not in testing.  I'll continue to keep an eye on it.

Edit: I also see by inspection that with -N2 the tests are definitely being run 2 at a time.

Edit 2:  It seems that older pipelines run the tests at double speed, but only half as many at once.  There is still a local speed up depending on the number of cores your machine has.  But this is overall disappointing...

Edit 3: tests are mysteriously failing on cabal 900 and not cabal 810!

Fizzixnerd committed:
Fix problem where subnames of previous tests would match on later ones

facundominguez commented on tests/test.hs:
```diff
@@ -53,6 +53,9 @@ data DependentTests = DependentTests
   }
 
+newtype SequentialTests = SequentialTests
+  { getTests :: [(TestTree, Int)] }
```
This could be a good place to document the meaning/purpose of the `Int` values stored in `getTests`.

Fizzixnerd replied:
I refactored this to just not store the Int.

facundominguez commented on tests/test.hs:
```diff
@@ -432,17 +435,37 @@ selfTests
+testSequentially name (SequentialTests indexedTests) =
+  let grouped = (\(t, n) -> (testGroup (name <> show n) [t], n)) <$> indexedTests
+      -- turns i.e. [a, b, c, d] into [(a, b), (b, c), (c, d)]
+      pairs = zip <*> tail $ grouped
```
I did struggle to figure this one.  How about
```
pairs = zip grouped (tail grouped)
```
?

Fizzixnerd replied:
Sorry, I always use this form because of https://stackoverflow.com/questions/3774247/what-do-we-call-this-new-higher-order-function/3775293#3775293.  I will change it!

facundominguez commented on tests/test.hs:
```diff
@@ -432,17 +435,37 @@ selfTests
+  in
+    case grouped of
+      [] -> []
+      (t:_) -> fst t : deps
```
Is this function similar in effect to
```
testSequentially name (SequentialTests indexedTests) =
    localOption (NumThreads 1) (testGroup "sequential" $ map fst indexedTests)
```
?

Fizzixnerd replied:
Oddly, no.  I tried this, and it ran on all my CPUs regardless.  However, doing `stack --nix test --ta '-j1 -p sequential'` DID work.  I do like the idea of adding `sequential` to the test name, however, so that you can run the sequential tests by name.

facundominguez commented on tests/test.hs:
```diff
@@ -439,17 +439,23 @@ testGroupsWithLibs name (DependentTests libTests nonlibTests) =
+  where
+    -- Supports up to 1000 tests -- anymore, and you'll get a dependency cycle
+    -- issue because "X-100" matches, say"X-1000"
+    mkName n = printf "%s-%03d" name n
```
I'd say you want to throw an error if `n` exceeds this limit instead of leaving it to tasty to fail.

That said, is it possible to write the pattern string so the match requires the full name?

Fizzixnerd replied:
I can write the `mkName` so that it appends a dot, i.e. `printf "%s-sequential-%d."`.... this should fix this issue.  Not sure why I didn't think of it before!

facundominguez commented on tests/test.hs:
```diff
@@ -414,29 +411,35 @@ textOrder = Just . mkOrder $
+testGroupsWithLibs :: String -> ([TestTree], [TestTree]) -> [TestTree]
+testGroupsWithLibs name (libTests, nonlibTests) =
+  [ testGroup (name <> "Lib") libTests
+  , after AllFinish (name <> "Lib") $ testGroup name nonlibTests
```
Would `AllSucceed` work better here?

Fizzixnerd replied:
Personally, I think we should retain the ability to run all the tests, even if previous ones fail, since only certain tests in `nonLibTests` will rely on the `Lib` tests, and it would be inconvenient to have the failure of one of the lib tests shadow the failure of other nonlib ones.

Fizzixnerd committed:
Address PR comments:

- replace zip`ap`tail
- don't store the Int in the SequentialTests
- have mkName generate names which cannot be subnames of later tests

Fizzixnerd committed:
Ignore Data/Text/Foreign.hs

Fizzixnerd commented:
I've added the failing test to the text ignore list... quoting from my slack post:

> Having trouble with CI in https://github.com/ucsd-progsys/liquidhaskell/pull/1925 : not sure why, but cabal_900 is failing on Data/Text/Foreign.hs.  I look at a previous working CI job, https://app.circleci.com/pipelines/github/ucsd-progsys/liquidhaskell/1178/workflows/cdc139d2-ee49-45d9-9b90-7d3cd24d9f6c/jobs/7426?invite=true#step-112-1302, and noticed this test is not even run here, but the test is not in the text-ignored list. https://github.com/ucsd-progsys/liquidhaskell/issues/1491 points out a similar issue someone was dealing with?  Also, the test name is very strange when given to "rerun it", mentioning typeclass-pos ... I'm tempted to add this to the ignore list for test, but I don't understand why I have to when it wasn't being run before!  Anyone have any ideas?

I have ignored this test because it was only being run on the 810 CircleCIs, and fails on the 900 one.  I still do not know why the test wasn't picked up before (but is now), which is a bit worrying... I'm not sure a better way to deal with it though, since it was already being ignored (somehow).

facundominguez commented on tests/test.hs:
```diff
@@ -443,19 +443,16 @@ testGroupsWithLibs name (DependentTests libTests nonlibTests) =
-      -- turns i.e. [a, b, c, d] into [(a, b), (b, c), (c, d)]
-      pairs = zip <*> tail $ grouped
+testSequentially name (SequentialTests tests) =
+  let grouped = (\(t, n) -> (testGroup (mkName n) [t], n)) <$> (zip tests [0 :: Int ..])
```
The need to create a `testGroup` is that there is no way to get the name of the given `TestTree`, right?
If that is the case, a comment informing this could be helpful if someone wants to change it later on.

Fizzixnerd replied:
This is correct, and a comment has been added.

facundominguez commented on tests/test.hs:
```diff
@@ -779,6 +779,7 @@ textIgnored :: [FilePath]
 textIgnored 
   = [ "Setup.lhs"
     -- , "Data/Text/Axioms.hs"
+    , "Data/Text/Foreign.hs"                -- Breaks on GHC 9.0.1
```
IIUC this disables the test in all configurations, not just GHC 9.0.1.

There is this PR that disabled this test in the CI configuration: https://github.com/ucsd-progsys/liquidhaskell/pull/1872

I guess the options are:
1. Use CPP macros to skip tests for ghc9, or
2. Update the CI configuration after the changes to the test name

In general, I prefer to avoid CPP, but you let me know if the current arrangement is too confusing.

Fizzixnerd replied:
Because 1) `test.hs` already makes use of CPP; 2) the fact there are two ignore lists for tests is highly confusing (to me, at least); and 3) the name of the test in question is generated and depends on the number of tests run before it, and so cannot be reliably managed outside of `test.hs`; I propose that CPP is the preferable answer in this case.

facundominguez replied:
Works for me.

Fizzixnerd committed:
Move test skipping logic to test.hs from circleci config

Fizzixnerd committed:
Document the reason we create a singleton testGroup

facundominguez reviewed (approved):
Thank you, Matt!
